File: chap5_DFT/hydrogen/testHydrogenPoisson/Numerov.m
function solution = Numerov(h, iStart, iEnd, F, uStart, uNext, solution)
% u''(r)=F(r)
% h: integration step
% iStart: index of the starting point
% iEnd: index of the ending point
% F: array of F
% uStart: initial value of u 
% uNext: next value of u
% solution: u

if h < 0
    iStep = -1;
else
    iStep = 1;
end

h2 = h^2;
factor = h2 / 12;

wPrev = (1-factor*F(iStart)) * uStart;
solution(iStart) = uStart;

u = uNext;
solution(iStart+iStep) = uNext;
w = (1-factor*F(iStart+iStep))*u;

for i=iStart+iStep:iStep:iEnd-iStep
    % i is the central index
    wNext = 2*w - wPrev + h2*u*F(i); % (2.12)
    wPrev = w;
    w = wNext;
    u = w / (1-factor*F(i+iStep));
    solution(i+iStep) = u;
end


end

File: Thijssen-Solutions/lib/src/numerov/numerov.f
       SUBROUTINE Numerov (Delta, StartI, EndI, MaxSol, FArr,
     .                     Sing, PhiStart, PhiNext, Solution)

C Integrates the Schrodinger equation. The initial values
C of the solution are PhiStart and PhiNext resectively. The integration
C step is Delta, and the integration steps run from StartI to EndI. StartI may be larger
C than EndI; in that case, integration is performed backward.
C The output values is the solution, stored in the array "Solution".
C Sing determines whether the potential contains a singularity at
C r=0. 
C If there is a singularity at r=0, the value of the Numerov 
C function w at r=0 is taken equal to PhiStart, and not 
C equal to PhiStart/(1-h^2 FArr/12).

C This array is declared with linear size MaxSol.
C Delta is the integration step.
C The equation solved is
C Psi''(R_I) = FArr(I) Psi(R_I)
C FArr must therefore be filled with the appropriate values before
C calling the present routine. In the case of the radial 
C Schrodinger equation, FArr would contain the values
C FArr(I) = 2*(V(R)-E)+L*(L+1)/R**2 for R=R_I.

       IMPLICIT NONE

       INTEGER I, StartI, EndI, MaxSol, IStep

       DOUBLE PRECISION Phi, PhiStart, PhiNext, W, WNext, WPrev,
     .        Delta, Fac, FArr(MaxSol), DeltaSq, Solution(MaxSol)

       LOGICAL Sing

       IF (Delta.LT.0) THEN
         IStep = -1
       ELSE
         IStep = 1
       ENDIF

       DeltaSq = Delta*Delta
       Fac = DeltaSq/12.D0

       IF (Sing) THEN
         WPrev = PhiStart
       ELSE
         WPrev = (1-Fac*FArr(StartI))*PhiStart
         Solution(StartI) = PhiStart
       ENDIF

       Phi = PhiNext
       Solution(StartI+IStep) = PhiNext
       W   = (1-Fac*FArr(StartI+IStep))*Phi

       DO I = StartI+IStep, EndI-IStep, IStep
          WNext = W*2.D0 - WPrev + DeltaSq*Phi*FArr(I)
          WPrev = W
          W     = WNext
          Phi   = W/(1-Fac*FArr(I+IStep))
          Solution(I+IStep) = Phi
       ENDDO
       END 


